### DISCUSSION TITLE
Separate texture SRC attribute

### DISCUSSION THREAD

**Discussion by Lemovision** [google/model-viewer#2751] [Q&A]
Hi

For my application (custom user designs) I'm using a single model stored on my server
And I need multiple viewer pages referring to this model, each with a different texture file
Now by default the viewer looks for the texture in the same directory as the model
Can we have a separate attribute for the source path of the texture please?

Here is an example of same model with separate texture reference with 3dviewer.net (folder 01 and folder 02)

https://3dviewer.net/embed.html#model=https://lemovision.com/elephant/assets/Elephant.glb,https://lemovision.com/elephant/assets/01/cbimage.png
https://3dviewer.net/embed.html#model=https://lemovision.com/elephant/assets/Elephant.glb,https://lemovision.com/elephant/assets/02/cbimage.png

(I saw you have a function to "swap textures" on the fly, but then I would have to load the texture twice so I suppose it would be slower than referring to the right texture in the first place)

**Comment by elalish** [comment]
No, because we also have `createTexture()`, so the best way is to make a version of your model without the texture, then load the one you  want on the fly: https://modelviewer.dev/examples/scenegraph/#createTexturesExample

**Reply by Lemovision** [reply]
Thanks for the reply
Actually I'm more of a designer than a programmer
This example and method is a bit hard for me to understand (that's why I thought a SRC would be user-friendly)
I trying to adapt the example to create the texture directly at start instead of the dropdown, but no success yet
Could you please have a more simple example of createTexture() without the dropdown selection?
Or maybe elaborate the documentation for these methods? (I think there is just this example and no doc yet)

**Reply by elalish** [reply]
Good call on adding the docs; @timmmeh can you put that on your list?

**Reply by Lemovision** [reply]
Okay I got it to work on start similar to the example,
without the stuff related to the dropdown selection

```
<script type="module">
const modelViewerTexture = document.querySelector("model-viewer#elephant");

modelViewerTexture.addEventListener("load", () => {

  let material = modelViewerTexture.model.materials[0];

  let createAndApplyTexture = async (channel, event) => {
    
      // Creates a new texture.
      let texture = await modelViewerTexture.createTexture("https://lemovision.com/elephant/assets/01/cbimage.png");
      // Applies the new texture to the specified channel.
      material[channel].setTexture(texture);
     
  }

    createAndApplyTexture("normalTexture", event);
  
});
</script>
```

So this load my file as a normal texture like the example right
But if I choose "baseColorTexture" instead of "normalTexture" it doesn't work
How to make that work?

`Uncaught (in promise) TypeError: Cannot read property 'setTexture' of undefined at createAndApplyTexture`

**Reply by elalish** [reply]
The problem is `baseColorTexture` is not a member of a glTF `material`, but `material.pbrMetallicRoughness`: https://modelviewer.dev/examples/scenegraph/#swapTextures

**Reply by Lemovision** [reply]
Also I noticed a problem with this createTexture() method() on startup
First the model shows up naked without any texture...
And then only when the texture is loaded it appears on the model

I guess this works as intended when swaping textures on the fly
But I don't think it works well as a start method
I think model and texture should appear together when both loaded

So how about that Texture SRC attribute to load together?
Wouldn't it be good and more user friendly?

```
<model-viewer
src="https://lemovision.com/elephant/assets/Elephant.glb"
tsrc="https://lemovision.com/elephant/assets/01/cbimage.png"
</model-viewer>
```

**Reply by elalish** [reply]
well, except that `tsrc`  isn't very specific; there are lots of texture slots on lots of materials, hence why we have a large JS API for this. You can accomplish what you want by setting `reveal = "manual"` and calling `dismissPoster()` when your async load has finished.

**Reply by Lemovision** [reply]
Oh okay thanks I gave it a try
I had to set up the loading on "eager" to make it work I think, is that right?
Now the other problem I'm seeing is that the loading bar at the top completes when the model only is loaded
And then while the texture is still loading, there's no visible loading bar going on and no interaction possible for the user
So it might be confusing for users with slow connection to be stuck on the poster

**Reply by elalish** [reply]
Yeah, that's a little tricky. You can put a spinner while you're fetching the texture. And you can  show the model without a texture if you want them to have a quicker interactive experience. All UX tradeoffs; would love to see what you come up with.

**Reply by Lemovision** [reply]
Tbh I had the same issue with your swap texture examples actually
When I clicked on the dropdown I thought it didn't work because nothing was apparently happening for a while
I only saw it working when I came back to it afterwards, because it was loaded then

Maybe you can update the examples too with a loading bar solution
And I'll just copy paste it because I've no idea honestly ^^'

**Reply by Lemovision** [reply]
So I just made two separate loading texts for now, so the user knows what's going on
Maybe I'll add some spinner later yes... although a loading progress would be nice
I looks like that, with the same model, and different test textures loaded

https://lemovision.com/elephant/view/?e=01
https://lemovision.com/elephant/view/?e=02

Actually it would be great if your createTexture() method would trigger the top loading bar again
Is it possible that you could link that, same as when loading the model?
On the doc I saw you're tracking a [progress](https://modelviewer.dev/docs/#entrydocs-loading-events-progress) for model, environment, and skybox
How about texture loading too?
